### PR TITLE
Speed up multiple attributes overwrite detection. Fixes #24467

### LINKED ISSUE #24467: Blazor WebAssembly perf: Optimize multiple-attributes overwrite detection
This is part of #22432

Currently, catch-all parameters (a.k.a. additional attributes) are disproportionately expensive because of the extra time we spend normalizing the list of parameters each time we pass them to child components. We walk through the set of unmatched attributes and build a dictionary so we can detect which ones were overwritten by others. This is particularly expensive because of the cost of hashing each of the attribute name strings.

As a faster alternative, we can use a different data structure (other than `Dictionary<string, int>`) that is more optimized for the case where most lookups will not be a match (i.e., most parameters aren't overwritten by others). In the prototype below, `SimpleStringIntDict` has an API like `Dictionary<string, int>`, but works without having to call `key.GetHashCode()`. Instead it provides its own extremely basic and fast hash function that doesn't consider all the characters in the string. Since in most cases each lookup results in "no match", there's no drawback (in those cases) in having such a simplistic hash. This would only perform worse than `Dictionary<string, int>` in extraordinary cases, e.g., there are a large number of different parameters that all clash on this simplistic hash, and they all get overwritten many times within the same invocation. I don't expect there to be any reasonable use case like that.

Prototype implementation is at https://github.com/dotnet/aspnetcore/commit/0c098595377c192a21545e9ead791fa675262b24. The default ComplexTable benchmark unfortunately doesn't make heavy use of additional attributes, but if it's changed to pass three additional attributes per `<Cell>`, this optimization improves its timings by 17%. So I regard this optimization as very valuable for cases where people do use additional attributes heavily.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Details in #24467

I spent ages on this because there were so many ways to approach it. The awkward thing is I can actually make this quite noticeably faster still than the code in this PR, by implementing:

 * A custom dictionary that's hardcoded to use these equality/hash rules (i.e., not `System.Collections.Generic.Dictionary<T,U>`, but [a different implementation altogether](https://github.com/dotnet/aspnetcore/pull/24561/commits/a8adf781d63fce3b78b78143525ba7c1708e4dc1#diff-473a957df9e0c00bec2fe4c0d6d93b92)). This gives about another 2.5% gain on the benchmark on top of this PR.
 * A custom implementation of `char.ToLowerInvariant` that special-cases ASCII, but falls back on `char.ToLowerInvariant` for other inputs. This gives another 1% gain.

However after some wrangling I've decided not to do either of those. It's because even if my implementations are faster on the benchmark scenarios today, it's possible that:

 * Other scenarios would have different characteristics. The behavior of `System.Collections.Generic.Dictionary` in all extreme cases is well understood, but we can't say the same for new alternatives.
 * As the runtime evolves, things like `char.ToLowerInvariant` can be expected to get faster, whereas custom implementations won't.

So, the code as-is in this PR improves intensive attribute splatting scenarios by around 5%. Not just the attribute-splatting part of the process, but the timings of the entire scenario that involves intensive attribute splatting. That's pretty worthwhile even if we forgo the extra ~3.5% we could get with a custom dictionary and char-normalizer.

### CODE REVIEW

File: src/Components/Components/src/Rendering/RenderTreeBuilder.cs [hunk 1]
```diff
@@ -707,41 +706,40 @@ internal void ProcessDuplicateAttributes(int first)
...
             // Now that we've found the last attribute, we can iterate backwards and process duplicates.
-            var seenAttributeNames = (_seenAttributeNames ??= new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase));
+            var seenAttributeNames = (_seenAttributeNames ??= new Dictionary<string, int>(SimplifiedStringHashComparer.Instance));
             for (var i = last; i >= first; i--)
             {
                 ref var frame = ref buffer[i];
                 Debug.Assert(frame.FrameTypeField == RenderTreeFrameType.Attribute, $"Frame type is {frame.FrameTypeField} at {i}");
 
-                if (!seenAttributeNames.TryGetValue(frame.AttributeNameField, out var index))
+                if (!seenAttributeNames.TryAdd(frame.AttributeNameField, i))
```
Comment (captainsafia) [inline src/Components/Components/src/Rendering/RenderTreeBuilder.cs:715]: 👍

File: src/Components/Components/src/Rendering/SimplifiedStringHashComparer.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            var keyLength = key.Length;
+            if (keyLength > 0)
+            {
+                // Consider just the length and middle and last characters.
+                // This will produce a distinct result for a sufficiently large
+                // proportion of attribute names.
+                return unchecked(
+                    char.ToLowerInvariant(key[keyLength - 1])
+                    + 31 * char.ToLowerInvariant(key[keyLength / 2])
```
Comment (captainsafia) [inline src/Components/Components/src/Rendering/SimplifiedStringHashComparer.cs:38]: Can we assign the coefficients here to constants?

Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/SimplifiedStringHashComparer.cs:38]: Generally I'd prefer not to add the extra abstraction unless it's a case where we think there are multiple uses that need to be kept in sync. Let me know if you want to discuss it - maybe there are other reasons I'm missing!

Comment (captainsafia) [reply src/Components/Components/src/Rendering/SimplifiedStringHashComparer.cs:38]: I posted the feedback in the interest of readability. I totally hear you about it not being technically necessary. Maybe the middleware group is an inline comment (e.g. `42 /* LIFE_THE_UNIVERSE_AND_EVERYTHING */`)

**Review by captainsafia — APPROVED** [review]
Nice change overall! Nice to see a few lines of code bring about a nice perf boost.

**Pull request merged**